[Substrate.io Migration] Runtime Upgrades (#19)
polkadot-developers/polkadot-docs: develop/parachain-devs/parachain-maintenance/runtime-upgrades.md

---
title: Runtime Upgrades
description: Explains how runtime versioning and storage migration support forkless upgrades for Polkadot SDK-based networks and they factor in to chain upgrades.
---

# Runtime Upgrades

## Introduction

> One of the defining features of Polkadot SDK-based blockchains is the ability to perform forkless runtime upgrades. Unlike traditional blockchains, which require hard forks and node coordination for upgrades, Polkadot networks enable seamless updates without network disruption. Forkless upgrades are achieved through the use of WebAssembly (Wasm) runtimes stored on-chain, which can be securely swapped in and upgraded as part of the blockchain's state. By leveraging decentralized consensus, runtime updates can be made trustlessly, ensuring continuous improvement and evolution without halting operations. This guide explores how runtime versioning, Wasm deployment, and storage migrations facilitate these upgrades.

Reviewer: ```suggestion
One of the defining features of Polkadot SDK-based blockchains is the ability to perform forkless runtime upgrades. Unlike traditional blockchains, which require hard forks and node coordination for upgrades, Polkadot networks enable seamless updates without network disruption.

Forkless upgrades are achieved through the use of WebAssembly (Wasm) runtimes stored on-chain, which can be securely swapped in and upgraded as part of the blockchain's state. By leveraging decentralized consensus, runtime updates can be made trustlessly, ensuring continuous improvement and evolution without halting operations.

This guide explores how runtime versioning, Wasm deployment, and storage migrations facilitate these upgrades.
```

> 

Reviewer: ```suggestion

## Process Overview

```

Reviewer: I don't know if that's the best title, but I feel like this doesn't flow in the introduction. So we can either create a section, or find a way to better embed into the intro

Reviewer 2: I reworked that section a little bit. wdyt?

---
title: Runtime Upgrades

> description: Explains how runtime versioning and storage migration support forkless upgrades for Polkadot SDK-based networks, and what important factors to consider when upgrading your chain.

Reviewer 3: The description has 178 characters. Should be between 120 and 160

---

Forkless runtime upgrades are a defining feature of the Polkadot SDK. The ability to update the runtime logic without forking the code base enables your blockchain to evolve and improve over time. This capability is made possible by including the definition of the runtime execution environment—the runtime WebAssembly blob—as an element in the blockchain's runtime state.

> !!!info "Why 'Forkless'?"
>     Traditional blockchains require a [hard fork](https://en.wikipedia.org/wiki/Fork_(blockchain){target=_blank}) when upgrading the state transition function of their chain.

Reviewer 3: ```suggestion
!!!info "Why 'Forkless'?"
    Traditional blockchains require a [hard fork](https://en.wikipedia.org/wiki/Fork_(blockchain){target=\_blank} when upgrading the state transition function of their chain.
```

    A hard fork requires all node operators to stop their nodes and manually upgrade to the latest executable.
    For distributed production networks, coordination of hard fork upgrades can be a complex process.

Because the runtime is part of the blockchain state, network maintainers can leverage the blockchain's capabilities for trustless, decentralized consensus to securely make enhancements to the runtime.

> In the FRAME, the [`system`](https://paritytech.github.io/polkadot-sdk/master/frame_system/index.html){target=_blank} pallet defines [the `set_code` extrinsic](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/enum.Call.html#variant.set_code){target=_blank} that is used to update the WebAssembly code for the runtime. 

Reviewer 3: ```suggestion
In FRAME, the [`system`](https://paritytech.github.io/polkadot-sdk/master/frame_system/index.html){target=_blank} pallet defines the [`set_code`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/enum.Call.html#variant.set_code){target=\_blank} extrinsic that is used to update the WebAssembly code for the runtime. 
```

!!!info "Parachain vs Solo Chain Upgrades"

>     While the `set_code` extrinsic is sufficient for updating the runtime in solo chain setups, parachains must undergo a different procedure of calling `authorize_upgrade`, followed by `apply_authorized_upgrade`. These extrinsics ensure that the relay chain is aware of the upgrade, and that it gets applied

Reviewer 3: ```suggestion
    While the `set_code` extrinsic is sufficient for updating the runtime in solo chain setups, parachains must undergo a different procedure of calling `authorize_upgrade`, followed by `apply_authorized_upgrade`. These extrinsics ensure that the relay chain is aware of the upgrade, and that it gets applied.
```

> When updating a runtime, there is a differentiation in _adding_ functionality to the runtime versus _updating_ existing functionality. In many cases, updating existing functionality, such as how a pallet stores state, would require a [storage migration](#storage-migration).

Reviewer 3: ```suggestion
When updating a runtime, there is a differentiation in _adding_ functionality to the runtime versus _updating_ existing functionality. In many cases, updating existing functionality, such as how a pallet stores state, would require a [storage migration](#storage-migrations).
```

## Runtime Versioning

The component that selects the runtime execution environment to communicate with is called the _executor_.  Although you can override the default execution strategies for custom scenarios, in most cases the executor select the appropriate binary to use by evaluating the following information for both the native and WebAssembly runtime binaries:

- `spec_name`
- `spec_version`

> - `authoring_version`

Reviewer 3: These three items are not introduced in the paragraph above. Maybe they need an explanation or can be deleted because they are mentioned in the table below

To provide this information to the executor process, the runtime includes a [runtime version struct](https://paritytech.github.io/polkadot-sdk/master/sp_version/struct.RuntimeVersion.html){target=_blank} similar to the following:

> ```rust
>     --8<-- 'code/develop/parachain-devs/parachain-maintenance/runtime_version.rs'
> ```

Reviewer 3: ```suggestion
```rust
--8<-- 'code/develop/parachain-devs/parachain-maintenance/runtime_version.rs'
```
```

| `transaction_version` | The version of the interface for handling transactions. This parameter can be useful to synchronize firmware updates for hardware wallets or other signing devices to verify that runtime transactions are valid. The parameter allows hardware wallets to know which transactions they can safely sign. This number must be bumped if there is a change in the index of the pallets in the `construct_runtime!` macro or if there are any changes to dispatchable functions, such as the number of parameters or parameter types. If this number is updated, then the `spec_version` must also be updated. |

> | `apis`                | A list of supported [runtime APIs](https://paritytech.github.io/polkadot-sdk/master/sp_api/macro.impl_runtime_apis.html){target=_blank} along with their versions.                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
> The executor that the native runtime has the same consensus-driven logic as the WebAssembly before it chooses to execute it. However, because the runtime versioning is set manually, the executor can still make inappropriate decisions if the runtime version is misrepresented.

Reviewer 3: ```suggestion
| `apis`                | A list of supported [runtime APIs](https://paritytech.github.io/polkadot-sdk/master/sp_api/macro.impl_runtime_apis.html){target=_blank} along with their versions.                                                                                                                                                                                                                                                                                                                                                                                                                                                            |

The executor that the native runtime has the same consensus-driven logic as the WebAssembly before it chooses to execute it. However, because the runtime versioning is set manually, the executor can still make inappropriate decisions if the runtime version is misrepresented.
```

> The executor that the native runtime has the same consensus-driven logic as the WebAssembly before it chooses to execute it. However, because the runtime versioning is set manually, the executor can still make inappropriate decisions if the runtime version is misrepresented.

Reviewer 3: ```suggestion
The executor follows the same consensus-driven logic for both the native runtime and the WebAssembly runtime before deciding which to execute. However, since runtime versioning is manually set, there is a risk that the executor could make incorrect decisions if the runtime version is misrepresented or incorrectly defined.
```

## Accessing the runtime version

The runtime version can be accessed through the `state.getRuntimeVersion` RPC endpoint. The endpoint accepts an optional block identifier. It can also be accessed through the runtime metadata to understand the APIs the runtime exposes and how to interact with these APIs.

> The runtime metadata should _only_ change when the chain's [runtime `spec_version`](https://paritytech.github.io/polkadot-sdk/master/sp_version/struct.RuntimeVersion.html#structfield.spec_version){target=_blank} changes.

Reviewer 3: ```suggestion
The runtime metadata should _only_ change when the chain's runtime [`spec_version`](https://paritytech.github.io/polkadot-sdk/master/sp_version/struct.RuntimeVersion.html#structfield.spec_version){target=\_blank} changes.
```

## Storage Migrations

[Storage migrations](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/frame_runtime_upgrades_and_migrations/index.html#migrations){target=_blank} are custom, one-time functions that allow you to update storage to adapt to changes in the runtime.

For example, if a runtime upgrade changes the data type used to represent user balances from an _unsigned_ integer to a _signed_ integer, the storage migration would read the existing value as an unsigned integer and write back an updated value that has been converted to a signed integer.

If you don't make these kinds of changes to how data is stored when needed, the runtime can't properly interpret the storage values to include in the runtime state and is likely to lead to undefined behavior.

### Storage migrations with FRAME

> FRAME storage migrations are implemented using the [`OnRuntimeUpgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.OnRuntimeUpgrade.html){target=_blank} trait. The `OnRuntimeUpgrade` trait specifies a single function, `on_runtime_upgrade`, that allows

Reviewer 3: ```suggestion
FRAME storage migrations are implemented using the [`OnRuntimeUpgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.OnRuntimeUpgrade.html){target=\_blank} trait. The `OnRuntimeUpgrade` trait specifies a single function, `on_runtime_upgrade`, that allows
```

---
title: Runtime Upgrades
description: Explains how runtime versioning and storage migration support forkless upgrades for Polkadot SDK-based networks and they factor in to chain upgrades.
---

# Runtime Upgrades

> The on-chain forkless runtime upgrades are one of the defining features of Polkadot SDK-based blockchains. The ability to update the runtime logic without forking the network enables your blockchain to evolve and improve seamlessly over time. This is made possible through the [Wasm](https://webassembly.org/){target=\_blank} powered runtimes that are stored in the blockchain state. When upgrading the blockchain network, the Wasm code containing the new runtime logic is uploaded on-chain and is swapped with the old code when the upgrade is enacted.

Reviewer: I think this intro could benefit from the [formula for intros](https://github.com/papermoonio/documentation-style-guide/blob/main/style-guide.md#introductions) defined in the style guide

We could remove "Why Forkless" from the info admonition and have that be part of describing the current standard.

Reviewer: There's a lot of content here, so some of it may need to be broken up 🤷‍♀️

When updating a runtime, there is a differentiation in adding functionality to the runtime versus updating existing functionality. In many cases, updating existing functionality, such as how a pallet stores state, would require a [storage migration](#storage-migrations).

## Runtime Versioning

The component that selects the runtime execution environment to communicate with is called the executor.  Although you can override the default execution strategies for custom scenarios, in most cases the executor selects the appropriate binary to use by evaluating and comparing key parameters from the native and Wasm runtime binaries.

To provide this parameter information to the executor process, the runtime includes a [runtime version struct](https://paritytech.github.io/polkadot-sdk/master/sp_version/struct.RuntimeVersion.html){target=\_blank} similar to the following:

```rust
--8<-- 'code/develop/parachain-devs/parachain-maintenance/runtime_version.rs'
```

The parameters in the struct provide the following information to the executor:

> - **`spec_name`** - The identifier for the different runtimes

Reviewer: ```suggestion
- **`spec_name`** - the identifier for the different runtimes
```

watch capitalization of the first word, it should be lowercase

- **`impl_name`** - The name of the implementation of the spec. This is of little consequence for the node and serves only to differentiate code of different implementation teams
- **`authoring_version`** - The version of the authorship interface. An authoring node won't attempt to author blocks unless this is equal to its native runtime
- **`spec_version`** - The version of the runtime specification

>     ??? info "Additional information"

Reviewer: Ah yeah, not a huge fan of these dropdowns. It sort of distracts you from the actual definition haha

Reviewer 2: Took them out ;-)
